messages.go

```diff
@@ -182,3 +182,126 @@ func NewMsg(msgType string, parent ComposedMsg) (ComposedMsg, error) {
+	return receipt.SendResponse(receipt.Sockets.IOPubSocket, msg)
+}
+
+// Reply creates a new ComposedMsg and sends it back to the return identities over the
+// Shell channel.
+func (receipt *msgReceipt) Reply(msgType string, content interface{}) error {
+	msg, err := NewMsg(msgType, receipt.Msg)
+
+	if err != nil {
+		return err
+	}
+
+	msg.Content = content
+	return receipt.SendResponse(receipt.Sockets.ShellSocket, msg)
+}
+
+// MIMEDataBundle holds data that can be presented in multiple formats. The keys are MIME types
+// and the values are the data formatted with respect to it's MIME type. All bundle should contain
+// at least a "text/plain" representation with a string value.
+type MIMEDataBundle map[string]interface{}
+
+// NewTextMIMEDataBundle creates a MIMEDataBundle that only contains a text representation described
+// the the parameter 'value'.
+func NewTextMIMEDataBundle(value string) MIMEDataBundle {
+	return MIMEDataBundle{
+		"text/plain": value,
+	}
+}
+
+type KernelStatus string
+
+const (
+	KernelStarting KernelStatus = "starting"
+	KernelBusy                  = "busy"
+	KernelIdle                  = "idle"
+)
+
+// PublishKernelStatus publishes a status message notifying front-ends of the state the kernel is in.
+func (receipt *msgReceipt) PublishKernelStatus(status KernelStatus) error {
+	return receipt.Publish("status",
+		struct {
+			ExecutionState KernelStatus `json:"execution_state"`
+		}{
+			ExecutionState: status,
+		},
+	)
+}
+
+// PublishExecutionInput publishes a status message notifying front-ends of what code is
+// currently being executed.
+func (receipt *msgReceipt) PublishExecutionInput(execCount int, code string) error {
+	return receipt.Publish("execute_input",
+		struct {
+			ExecCount int    `json:"execution_count"`
+			Code      string `json:"code"`
+		}{
+			ExecCount: execCount,
+			Code:      code,
+		},
+	)
+}
+
+// PublishExecuteResult publishes the result of the `execCount` execution as a string.
+func (receipt *msgReceipt) PublishExecutionResult(execCount int, output string) error {
+	return receipt.Publish("execute_result",
+		struct {
+			ExecCount int            `json:"execution_count"`
+			Data      MIMEDataBundle `json:"data"`
+			Metadata  MIMEDataBundle `json:"metadata"`
+		}{
+			ExecCount: execCount,
+			Data:      NewTextMIMEDataBundle(output),
+			Metadata:  make(MIMEDataBundle),
+		},
+	)
+}
+
+// PublishExecuteResult publishes a serialized error that was encountered during execution.
+func (receipt *msgReceipt) PublishExecutionError(err string, trace []string) error {
+	return receipt.Publish("error",
+		struct {
+			Name  string   `json:"ename"`
+			Value string   `json:"evalue"`
+			Trace []string `json:"traceback"`
+		}{
+			Name:  "ERROR",
+			Value: err,
+			Trace: trace,
+		},
+	)
+}
+
+type Stream string
+
+const (
+	StreamStdout Stream = "stdout"
+	StreamStderr        = "stderr"
+)
+
+// PublishWriteStream prints the data string to a stream on the front-end. This is
```

Comment:
Is this function or the above constants and types used at this point?  If not, let's remove and add them when needed.